### DISCUSSION TITLE
Launching Logtalk with common loading flag

### DISCUSSION THREAD

**Discussion by PaulBrownMagic** [LogtalkDotOrg/logtalk3#93] [Ideas]
Different Prolog systems accept an initial file to consult in different manners, some with a flag and some just as the first arg.

It'd be nice to create a uniform interface for the Logtalk integration scripts (eclipselgt, gplgt, swilgt, etc) such that we can always specify the file to `logtalk_load/1`. This would allow us to launch Logtalk like so (if the `-l` flag were chosen for the implementation):

```bash
:~$ logtalk -l loader.lgt
```
This should always be the same no matter which backend was selected.

**Comment by pmoura** [comment]
That's an idea that surfaced in the past. As you likely know, there's no standard set of command-line options shared by most Prolog systems. Thus, the idea of the Logtalk integration scripts abstracting a subset of those options (notably, loading a file and calling a goal) in order to present a common view to the user is certainly appealing. On the other hand, users may also expect to be able to (as it's possible with the current scripts) use the specific command-line options provided by a Prolog system (the integration scripts simply pass those options to the backend). One possible solution would be to use long options names, thus both avoiding conflicts with backend options while still allowing them to be used in tandem.

**Comment by pmoura** [comment]
Found a StackOverflow discussion on how to handle long options using the Bash built-in `getopts` function:

https://stackoverflow.com/questions/402377/using-getopts-to-process-long-and-short-command-line-options#7948533

This seems to provide an implementation solution without the portability issues (across operating-systems) of the `getopt` alternative.

**Reply by PaulBrownMagic** [reply]
Looking into it `getopts` only supports short-flags, but is more cross-platform as a bash built-in.  Whereas `getopt` supports those long flags, but is "third-party" and popular, so likely to be in bash but not guaranteed. Looks like if we want to take a flag for the logtalk loader we'd be looking for a short-flag that hasn't been taken.

Alternatively, perhaps the adapter script could simply take the loader file as a positional argument? Usage would be:

```
1) logtalk [options]
2) logtalk loader_file [options]
```

Where `[options]` are passed to the backend. We could use bash to test if `$1` starts with a `-` so we'll know if we're in usage case 1 or 2. Here's a short test script:

```
if [[ $1 != -* ]]; then
	LOADER_FILE=$1
	shift
fi
echo $LOADER_FILE
echo $@
```
That assigns the first argument to `$LOADER_FILE` if it doesn't start with `-` and then shifts `$@` to remove the first argument from it. Otherwise `$@` is left untouched. This could be used with each integration file so when you call whatever goal loads the integration prolog file it could be followed by `{$LOADER_FILE}`.

What do you think of it being positional?

**Reply by pmoura** [reply]
One of the answers to that StackOverflow question shows how to use the `getopts` function to parse long options. But I'm yet to play with it.

**Reply by pmoura** [reply]
One issue I see with assuming that `$1` is the name of a Logtalk source file is that several backends also support loading Prolog files in the same way. We could check for a Logtalk file name extension, however (`.lgt` or `.logtalk`).

**Reply by PaulBrownMagic** [reply]
Every integration script with the exception of Scryer puts a flag before `$@`, which already breaks this first-arg source file behaviour. In the case of Scryer maybe we could change it to:

```
exec scryer-prolog "$LOGTALKHOME/integration/logtalk_scryer.pl" "$@"
```
And then we don't need to process the args for it at all? Or does it even accept command-line flags? I can't see that it does and so we wouldn't be able to pass a loader in anyway?

Then to adapt the above script to check for a logtalk file rather than a flag:
```
if [[ $1 == *.lgt ]] || [[ $1 == *.logtalk ]]; then
	LOADER_FILE=$1
	shift
fi
echo $LOADER_FILE
echo $@
```
With that I think we're reconstructing the first-arg source file where it is already broken.

**Reply by pmoura** [reply]
Scryer Prolog doesn't currently provide a a command-line option for loading a file, instead requiring:

```bash
$ scryer-prolog -h
Usage: scryer-prolog [OPTIONS] [FILES] [-- ARGUMENTS]

Options:
   -h, --help           Display this message
   -v, --version        Print version information and exit
   -g, --goal GOAL      Run the query GOAL
```

**Reply by PaulBrownMagic** [reply]
Ah, so we can keep it as it is, and if we have a `$LOADER_FILE` send the load as a goal.

First-arg loader is looking good...

Want me to do SWI, GNU, ECLiPSe, and LVM example integration scripts?

**Reply by pmoura** [reply]
For SWI-Prolog, it would be:

```bash
if [[ $1 == *.lgt ]] || [[ $1 == *.logtalk ]]; then
	LOADER_FILE=$1
	shift
	exec swipl -s "$LOGTALKHOME/integration/logtalk_swi.pl" -g "logtalk_load('$LOADER_FILE')" "$@"
else
	exec swipl -s "$LOGTALKHOME/integration/logtalk_swi.pl" "$@"
fi
```

Similar for other backends.

**Reply by PaulBrownMagic** [reply]
Yes, alternatively:

```
if [[ $1 == *.lgt ]] || [[ $1 == *.logtalk ]]; then
	LOADER_FILE="-g logtalk_load('$1')"
	shift
fi
exec swilgt -s "$LOGTALKHOME/integration/logtalk_swi.pl" $LOADER_FILE "$@"
```
(No quotes for `$LOADERFILE`)

That'll keep a single exec line then.

**Reply by pmoura** [reply]
Quotes may still be required for the goal itself.

**Reply by pmoura** [reply]
Implemented the necessary changes but testing exposed issues with several backends, including CxProlog, ECLiPSe, SICStus Prolog, and YAP. These systems don't  support multiple goal options. Thus, supporting loading a Logtalk source file by simply passing its name precludes also passing a goal. To be clear, concatenating goals at the integration script level as a workaround is possible but undesirable complexity.

**Reply by PaulBrownMagic** [reply]
Well that's annoying, I agree that's a bit much complexity to be enjoyable in bash.

Another idea: we let Logtalk handle the flag rather than bash. In SWI, GNU and Cx (I'm installing more dialects! But YAP failed to build...), ECLiPSe and LVM we can use `--` to distinguish flags passed to the program. These backends support `current_prolog_flag(argv, Args)`, except ECLiPSe which gives us `argc/1` and `argv/2` instead. I get the impression `current_prolog_flag(argv, _)` is a defacto-standard?

So how about `logtalk [options to backend] -- loader.lgt [plus other flags to argv]` and then in the integration prolog scripts we see if there's a loader file in the argv and if there is, we load it? And if we want a long flag: `logtalk [options] -- --loader_file loader.lgt`

With this we don't mess with any backends' flags, we only prohibit the user from using the long flag: `--loader_file` (and/or maybe a short-flag `-l`) for their own program and order after the `--` is also not important.

**Reply by pmoura** [reply]
You can check how different backends handle (if at all!) `--` in the `os` library. But `--` is important as it allows passing arbitrary arguments to an application and to tests at the command-line.

The only way out of this portability mess is, I think, long options that would be Logtalk specific.

**Comment by PaulBrownMagic** [comment]
Long flags makes a lot of sense, good idea!
